utils/docs/refdes_renum.1.in
.TH refdes_renum 1 "@DATE@" "gEDA Project" @VERSION@

.SH NAME
refdes_renum \- gEDA/gaf refdes renumbering utility
.SH SYNOPSIS
.B refdes_renum [\-\-nocopy] [\-\-pgskip [
.I number
.B ] ]
.I file1
.B ...
.PP
.B refdes_renum
.BR \-\-help | \-h
.PP
.B refdes_renum \-\-version
.SH DESCRIPTION
.B refdes_renum
reads a gschem schematic file or files and renumbers all reference
designators.  The reference designators are numbered starting with 1 and the
old schematic file is replaced by the modified schematic file.
.SH OPTIONS
.TP
.BR \-\-help ", " \-h
Display a help message.
.TP
.B \-\-nocopy
If given, this flag leaves the modified files in new files
whose names are generated by appending a `.renum' to the
original file names.  The default is to overwrite the original.
.TP
.B \-\-pgskip
When this flag is used, components on the first schematic sheet
are numbered starting with 101.  On the second sheet, they start
with 201, etc  Specifying a value gives the step between pages.
For example \-\-pgskip 10 will start with 11, 21, 31, etc.
.TP
.B \-\-gentle
This flag tells refdes_renum to leave any refdeses
alone if they already have numbers.  Use this option to number
new components in a schematic which has already been numbered.
Note that \-\-gentle is set by default!
.TP
.B \-\-force
Set this flag to renumber all refdeses, whether they are already
numbered or not.
.TP
.B \-\-verbose
Enables verbose output.
.TP
.B \-\-version
Shows the version of this program.
.SH RETURN CODES
.B refdes_renum
returns the following codes to the shell upon completion:
.TP
0
Program ran successfully.
.TP
1
Error opening or reading input file.
.TP
2
Error opening or writing output file.
.TP
3
Too many components for \-\-pgskip setting.
.TP
4
Internal error (program bug encountered).
.SH EXAMPLES
To renumber the reference designators in a single page schematic mysch.sch:
.IP
refdes_renum mysch.sch
.PP
To renumber the reference designators in schematic pages pg1.sch, pg2.sch and
pg3.sch, with the reference designators on pg1.sch beginning at 101, the
designators on pg2.sch beginning at 201, and the designators on pg3.sch
beginning at 301:
.IP
refdes_renum \-\-pgskip pg1.sch pg2.sch pg3.sch
.SH AUTHOR
Dan McMahill <danmcmahill@netbsd.org>
.SH SEE ALSO
.BR grenum (1)
.SH COPYRIGHT
.nf
Copyright \(co 1999-@YEAR@ gEDA Contributors.  License GPLv2+: GNU GPL
version 2 or later.  Please see the `COPYING' file included with this
program for full details.
.PP
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2015 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf 2.1.0.0 - 04/01/15 Rampage Edition

* This release of the Rampage Edition of the gEDA-gaf suite commemorates
  the 17th Anniversary of the first public release of gEDA-gaf. The 2.09
  series releases reflected a very aggressive development of new features
  and major enhancements in many areas. Numerous enhancements focused on
  existing features and functions that did not always work as expected.

* New editing functions provide powerful methods to improve productivity.

  A new Extend Objects functions allows linear objects to be extended to
  other objects with a simple key press.

  A new Break function allows Arcs, Boxes, Circles, Linear objects, and
  Paths, both closed and open, to be broken at user specified points.

  A Powerful new Array dialog has been added to create arrays of selected
  objects. Creating Arrays allows both verb-noun or noun-verb operations.

  A new Auto-Snap Objects to Grid editing function has been added allowing
  Selections to be quickly repositioned to the current Grid snap. The Auto
  Snap features works with all objects types, including graphical and image
  objects.

  A new bimodal Offset function provides editing methods to quickly create
  new objects at programmable distances. The distance can be specified by
  entering a value in a dialog or offsetting Selection sets to the current
  pointer position using hot-keys. The offset distance is retained between
  operations.

  A new Select Last function has been added to select the last object created
  or edited.

  The visibility of netname attributes can now be toggled while inside an
  action.

  A new RC keyword; third-button-cancel, allows user to specify a preference
  for canceling operations using the third mouse button. In addition, object
  adder modes can now be cancel by double-clicking the final entry point.

  A new Zoom to Magnification feature allows user to zoom to a specified level.

* Picture replacement operations can now be undone.

* A new Export functions allow exporting of embedded objects, including image
  objects.

* Picture objects no longer need to maintain the original aspect ratio. Aspect
  ratios of image objects can be freely modified using Grips. The Ctrl key can
  be combined to maintain the aspect ratio identically to the old behavior.
  When exporting images using the new export feature, options are provided to
  export the image at the original or the current aspect ratio and size.

* Coordinates can now be displayed in the status bar using Cartesian and vector
  formats. The format can be changed at any time, include inside of an action,
  by right clicking on the coordinates widget and selecting the format from a
  pop-up menu. Vector formats revert to the previous selected format when not
  applicable to the active context.

* Pointer button assignments can now be set via a pop-up menu by clicking on
  the associated status widget on the Status bar. Assignments can still be set
  using in the Preferences Dialog or over-ridden using RC keywords, but the
  new method allows changing assignments while inside an action.

* The status text now changes to blue when inside of an editing action.

* F-Keys have been assigned to common/standard program functions.

* Menus in Gschem have been revised for better organization with enhanced
  options. The tools menu now includes access to Version tracking and
  alternative simulations back-ends. A toggle widget has been added to
  the Options menu to toggle a new drag-can-move preference variable in
  order to enable and disable dragging moving.

  A new File/Export main menu fly-out was added for options to export both
  embedded components and pictures. Exporting pictures utilized a new custom
  Image Chooser Dialog allowing user to select from a variety of options.
  All available image formats suppported by host libraries are supported.

  Draw-Order functions have been added to the Page menu as a fly-out using
  Scheme, allowing users to manipulate the order in which objects are drawn.

  The pop-up context menu now provides options for Copy, Move, Rotate, Mirror,
  Component Properties, Zoom to Mag, and Zoom to Selection. Additional options
  have also been add as fly-out side menus to the pop-up menu. Fly-out options
  use the X-Y location of the pointer when the context was invoked. Copy, Move,
  Rotate, and Mirror actions were reinstated and work identically to previous
  implementations.

  Additional icons are provided for "Move", Simulations, DRC, and attribute-
  reset. The default icon for "Move" was changed to have a more consistence
  theme with other editing icons.

  Menu tool-tips for File/Recent and Sessions/Restore now respond properly to
  user preferences. Tool-tips for symbol names on the Component dialog field
  shows the full path to symbol file.

  Anti-alias options were added to the Preference dialog in Gschem for the
  anti-alias method to be used by the chosen render. Modifing the anti-alias
  method longer requires a re-start for the new setting to take effect.

* Page Select Dialog has been enhanced to allow displaying file names without
  the paths, a toggle switch allows users to switch the path portion on or off
  in the tree-view. The user perference is retained between sessions.

* The appearance of Auto-Number Dialog has been improved and is more robust.
  The dialog was enhanced with frames and separate filter options for scope
  text including options for wild-cards and a list of recently used filters.
  Recent filters are not retained between sessions.

* The Line Type Dialog now includes an options combo to set the line end-type.


utils/scripts/refdes_renum.pl
#!/usr/bin/perl -w
#
# Copyright (C) 2003, 2012 Dan McMahill
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if  not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA


# This script is used to renumber devices in a gschem (part of gEDA)
# schematic.  This program works on single as well as multi-sheet
# schematics.
#
# Usage:
#    renum.pl file1.sch [file2.sch [file3.sch ...]]
#

# for parsing input options
use Getopt::Long;

# for ceil function
use POSIX;

# don't allow -he to be interpreted as --help
$Getopt::Long::autoabbrev=0;

#  my $clear; # reset all refdes
&GetOptions(("help|h" => \&usage,
	     "nocopy" => \$nocopy,
	     "pgskip:100" => \$pgskip,
	     "verbose" => \$verbose,
	     "version" => \&version,
             "clear"   => \$clear,
	     "gentle" => \$gentle,
	     "force" => \$force,
	     ));

usage() if $Getopt::Long::error;


# By default pgskip is empty.  But if the user gives
# the --pgskip option with no value, then assign pgskip
# to 100.
if (defined($pgskip) && ($pgskip == 0)) {
    $pgskip = 100;
}


# This makes --gentle the default behavior.  To
# force refdes renumbering, use --force.
$gentle = 1;
if ( defined($force) || defined($clear) ) {
    undef($gentle);
}

# Print usage string in event user didn't call out any args
usage() unless @ARGV;

# Make sure the input schematic exists and we can read it.
# Also count number of files to open.
$i=0;
while(@ARGV) {
  $fname[$i]=shift(@ARGV);
  die_sdb(1, "Schematic file $fname[$i] does not exist or can not be read")
    unless -r $fname[$i];
  $i++;
}
$filecnt = $i;  # number of files to open
if($verbose) {print "Found $filecnt files to process\n";}

# To refdes an entire design, we normally just loop through all files
# and process each one in order.  However if --gentle option is set,
# we need to make two passes through all files.  First to create a hash
# of the highest refdeses used, and second to update unnumbered
# refdeses.  The hash is used to know where to start the refdes numbering.
#
# It gets more complicated.  If --pgskip is not set, then we only need
# to find the highest refdes overall.  But if --pgskip is set, then
# we need to know the highest refdes on each *page*.  This is accomplished
# on the first scan by creating the variable $highdev{$i, $pre} when --pgskip
# is defined, but using the variable $highdev{$pre} when it is not.
#
if (defined($gentle)) {  # First scan through design to build hash
    for($i=0; $i < $filecnt; $i++) {   # $i is the file index.
	print "Scanning input file #".($i+1).": $fname[$i]\n";
	open(NETLIST,"$fname[$i]") or die_sdb(1, "Can't open $fname[$i]: $!\n");

	while($line = <NETLIST>) {  # Iterate over lines found in .sch file.
	    unless( $line =~ /^refdes/) {
		next;
	    }
	    # Found refdes line.  Eat the newline.
	    $line =~ s/\n//;

	    # Now extract component prefix and suffix.
	    if($verbose) {print ("Matching $line ....\n");}
	    $line =~ /refdes=([a-zA-Z_]+)(\d+|\??)/i;
	    my $pre = $1;
	    my $suf = $2 ? $2 : "?";
	    if($verbose) {print "Refdes line \"$line\" has pre=$pre suf=$suf\n";}

	    # Now put highest refdes found into $highdev hash.
	    if (defined($pgskip)) {
		if ($suf eq "?") {
		    # Part has no pre-existing refdes.
		    # Skip to next part.
		    next;
		} elsif (!defined($highdev{$i, $pre})) {
                    # refdes is not cached, but refdes exists on part.
		    # Create new cache entry
		    if ($suf < $pgskip) {
			# if pgskip=100, on pg 1, and suf = 23, set highdev=123.
			$highdev{$i, $pre} = ($i+1)*$pgskip + $suf;
		    } elsif ( ( ($i+1)*$pgskip < $suf) &&
			      ( ($i+2)*$pgskip > $suf) ) {
			# if pgskip=100, we're on pg 2, and suf = 204, set highdev=204.
			$highdev{$i, $pre} = $suf;
		    } else {
			# I don't know what to do!
			die_sdb(3, "Encountered refdes incompatable with --pgskip setting.  Refdes=$pre$suf.\n");
		    }

		} elsif ($highdev{$i, $pre} < $suf) {
		    # part has refdes and refdes is higher than cached
                    # value.  Store this new value as highest value.
		    $highdev{$i, $pre} = $suf;
		} elsif ($highdev{$i, $pre} >= $suf) {
		    # part has refdes, but refdes is lower than cached
                    # high value.  Leave it alone.
		    next;
		} else {
		    # Should never get here
		    die_sdb(4, "Invalid refdes with --pgskip set!  Exiting....\n");
		}

		if($verbose) {
		    print "On page ".($i+1).", caching highest refdes $pre$highdev{$i, $pre}\n";
		}
	    } else {
		if ($suf eq "?") {
		    # Part has no pre-existing refdes.
		    next;
		} elsif (!defined($highdev{$pre}) ) {
                    # refdes is not cached, and refdes exists on part.  Create new
                    # cache entry
		    $highdev{$pre} = $suf;
		} elsif ($highdev{$pre} < $suf) {
		    # part has refdes and refdes is higher than cached
                    # value.  Store this new value as highest value.
		    $highdev{$pre} = $suf;
		} elsif ($highdev{$pre} >= $suf) {
		    # part has refdes, but refdes is lower than cached
                    # high value.  Leave it alone.
		    next;
		} else {
		    # Should never get here
		    die_sdb(4, "Invalid refdes!  Exiting....\n");
		}
		if($verbose) {
		    print "Caching highest refdes $pre$highdev{$pre}\n";
		}
	    }

	}  # while($line = <NETLIST>)
	close(NETLIST);
    }  # for($i=0; $i < $filecnt; $i++)
} # if (defined($gentle))


# OK, now we can read through the netlist file again, assign refdeses,
# and write the output file.
for($i=0; $i < $filecnt; $i++) {   # $i is the file index.
    print "Now processing input file #".($i+1).": $fname[$i]\n";
    open(NETLIST,"$fname[$i]") or die_sdb(1, "Can't open $fname[$i]: $!\n");

    # open output netlist
    $outfname="$fname[$i].renum";
    open(OUTNET,">$outfname") or die_sdb(2, "Can't open $outfname: $!\n");

    # Iterate over lines found in .sch file.
    while($line = <NETLIST>) {
	unless( $line =~ /^refdes/) {
	    print OUTNET $line;
	    next;
	}

	# Found refdes line.  Eat the newline.
	$line =~ s/\n//;

	# Now extract component prefix and suffix.
	if($verbose) {print ("Processing $line ....\n");}
	$line =~ /refdes=([a-zA-Z_]+)(\d+|\??)/i;
	my $pre = $1;
	my $suf = $2 ? $2 : "?";
	if($verbose) {print "Refdes line \"$line\" has pre=$pre suf=$suf\n";}

	# Now finally update refdes
	if ($clear) {
	    # Just overwrite all refdeses upon clear
	    if($verbose) {print ("Clearing refdes=$pre$suf\n");}
	    print OUTNET "refdes=$pre?\n";
	} elsif (defined($pgskip) && defined($gentle)) {
	    # If highdev exists, then start devcnt there, otherwise, start
	    # devcnt at correct value for each page.
	    if (!defined($devcnt{$i, $pre}) && defined($highdev{$i, $pre})) {
		$devcnt{$i, $pre} = $highdev{$i, $pre};
	    } elsif (!defined($devcnt{$i, $pre}) && !defined($highdev{$i, $pre})) {
		$devcnt{$i, $pre} = ($i+1)*$pgskip ;
	    }

	    if ($suf eq "?") {
		$devcnt{$i, $pre}++;
		if ($devcnt{$i, $pre} >= ($i+2)*$pgskip) {
		    print STDERR "ERROR:  You are numbering more than $pgskip elements with\n";
		    print STDERR "prefix $pre on this sheet.  You will need to either\n";
		    print STDERR "reduce this number or specify a larger step with the\n";
		    print STDERR "--pgskip argument.\n";
		    die_sdb(3, "");
		}
		print "Renumbering $line to $pre$devcnt{$i, $pre}\n" if($verbose);
		print OUTNET "refdes=$pre$devcnt{$i, $pre}\n";
	    } else {
		print "Leaving line=$line alone\n" if($verbose);
		print OUTNET "refdes=$pre$suf\n";
	    }

	} elsif (defined($pgskip) && !defined($gentle)) {
	    if (!defined($devcnt{$i, $pre})) {
		$devcnt{$i, $pre} = ($i+1)*$pgskip ;
	    }
	    $devcnt{$i, $pre}++;
	    if ($devcnt{$i, $pre} >= ($i+2)*$pgskip) {
		print STDERR "ERROR:  You are numbering more than $pgskip elements with\n";
		print STDERR "prefix $pre on this sheet.  You will need to either\n";
		print STDERR "reduce this number or specify a larger step with the\n";
		print STDERR "--pgskip argument.\n";
		die_sdb(3, "");
	    }
	    print "Renumbering $line to $pre$devcnt{$i, $pre}\n" if($verbose);
	    print OUTNET "refdes=$pre$devcnt{$i, $pre}\n";

	} elsif (!defined($pgskip) && defined($gentle)) {
	    if (!defined($devcnt{$pre}) && defined($highdev{$pre})) {
		$devcnt{$pre} = $highdev{$pre};
	    } elsif (!defined($devcnt{$pre}) && !defined($highdev{$pre})) {
		$devcnt{$pre} = 0;
	    }
	    if ($suf eq "?") {
		$devcnt{$pre}++;
		print "Renumbering $line to $pre$devcnt{$pre}\n" if($verbose);
		print OUTNET "refdes=$pre$devcnt{$pre}\n";
	    } else {
		print "Leaving line=$line alone\n" if($verbose);
		print OUTNET "refdes=$pre$suf\n";
	    }
	} elsif (!defined($pgskip) && !defined($gentle)) {
	    if (!defined($devcnt{$pre})) {
		$devcnt{$pre} = 0 ;
	    }
	    $devcnt{$pre}++;
	    print "Renumbering $line to $pre$devcnt{$pre}\n" if($verbose);
	    print OUTNET "refdes=$pre$devcnt{$pre}\n";

	} else {
	    die_sdb(4, "Invalid state when trying to update refdes!  Exiting.....\n");
	}
    }  # while($line = <NETLIST>)
    close(NETLIST);
    close(OUTNET);


    # make this an option to not overwrite the original
    if( $nocopy ) {
	print "Leaving page #",$i+1," output in $outfname\n";
	}
    else {
	system("mv $outfname $fname[$i]");
    }
}

exit(0);


#######################################################################
#
# Subroutines
#
#######################################################################

#---------------------------------
# usage()
#
# prints program usage
#---------------------------------

sub usage {
  my $pname = $0;
  $pname =~ s/.*\///g;

  print "Usage:\n\n";
  print "\t$pname [--nocopy] [--pgskip [number] ] file1 [file2 [file3 ... ] ]\n";
  print "\t$pname --help | -h\n";
  print "\t$pname --version\n";
  print "\n";
  print "$pname reads a gschem schematic file or files and renumbers all reference\n";
  print "designators.  The reference designators are numbered starting with 1 and the\n";
  print "old schematic file is replaced by the modified schematic file.\n";
  print "\n";
  print "$pname accepts the following options:\n";
  print "\n";
  print "    --help, -h  Displays this help message.\n";
  print "\n";
  print "    --nocopy    If given, this flag leaves the modified files in new files\n";
  print "                whose names are generated by appending a \".renum\" to the\n";
  print "                original file names.  The default is to overwrite the original.\n";
  print "\n";
  print "    --pgskip    When this flag is used, components on the first schematic sheet\n";
  print "                are numbered starting with 101.  On the second sheet, they start\n";
  print "                with 201, etc  Specifying a value gives the step between pages.\n";
  print "                For example --pgskip 10 will start with 11, 21, 31, etc.\n";
  print "\n";
  print "    --gentle    This flag tells refdes_renum to leave any refdeses\n";
  print "                alone if they already have numbers.  Use this option to number\n";
  print "                new components in a schematic which has already been numbered.\n";
  print "                Note that --gentle is set by default!\n";
  print "\n";
  print "    --force     Set this flag to renumber all refdeses, whether they are already\n";
  print "                numbered or not.\n";
  print "\n";
  print "    --verbose   Enables verbose output.\n";
  print "\n";
  print "    --version   Shows the version of this program.\n";
  print "\n\n";
  print "Return codes:\n\n";
  print "\t$pname returns the following codes to the shell upon completion:\n";
  print "\n";
  print "          0   Program ran successfully.\n";
  print "          1   Error opening or reading input file.\n";
  print "          2   Error opening or writing output file.\n";
  print "          3   Too many components for --pgskip setting.\n";
  print "          4   Internal error (program bug encountered).\n";
  print "\n";
  print "Usage examples:\n\n";
  print "\tTo renumber the reference designators in a single page schematic\n";
  print "\tmysch.sch:\n";
  print "\n";
  print "\t\t$pname mysch.sch\n";
  print "\n\n";
  print "\tTo renumber the reference designators in schematic pages pg1.sch,\n";
  print "\tpg2.sch and pg3.sch, with the reference designators on pg1.sch\n";
  print "\tbeginning at 101, the designators on pg2.sch beginning at 201,\n";
  print "\tand the designators on pg3.sch beginning at 301:\n";
  print "\n";
  print "\t\t$pname --pgskip pg1.sch pg2.sch pg3.sch\n";
  print "\n\n";
  print "$pname was written by Dan McMahill <dmcmahill\@netbsd.org>\n";
  print "\n\n";
  exit(0);
}

#---------------------------------
# version()
#
# prints program version
#---------------------------------

sub version {
  open(PROG,"$0") or die_sdb(4, "Could not open \"$0\" to find version\n\n");
  my $pname = $0;
  $pname =~ s/.*\///g;

  while($line = <PROG>) {
    if( $line =~ /^#\s*\$Id.*\$/) {
      @words = split ' ',,$line;
      $version = $words[3];
      $date = $words[4];
      print "$pname ($0):  Version $version, $date\n";
      exit(0);
    }
  }
  die_sdb(4, "Could not determine version of \"$0\"\n\n");
}

#-------------------------------------------
# die_sdb(rc, string)
#
# Local version of die.  Sets return code
# for shell, then calls die(string)
#--------------------------------------------
sub die_sdb {
    my $rc = $_[0];
    my $string = $_[1];
    $! = $rc;
    die($string);
}

# ----------------------------------------------
#
# Change Log
#
# $Log$
# Revision 1.8  2007-04-17 20:19:23  pcjc2
# Merge changes from noscreen branch
#
# Revision 1.2.6.2  2007/04/17 16:19:01  pcjc2
# Sync with trunk
#
# Revision 1.7  2007/04/15 23:20:31  sdb
# Made --gentle the default behavior of refdes_renum.
#
# Revision 1.7  2007/04/15 SDB
# Made --gentle default behavior.  Now user must use --force
# to get old behavior, which renumbers all refdeses whether
# already numbered or not.
#
# Revision 1.6  2007/04/15 00:42:22  sdb
# Added --gentle flag to refdes_renum, which won't renumber refdeses
# which already have a number.
#
# Revision 1.5  4.9.2007 sdb.
# Added --gentle flag, and include logic to handle case
# where user mixes --gentle with --pgskip.
#
# Revision 1.4  2007/02/24 18:43:17  pcjc2
# Merge changes to date from noscreen branch.
#
# Revision 1.2.6.1  2007/02/11 23:59:10  pcjc2
# Sync with trunc
#
# Revision 1.3  2007/02/10 20:46:17  sdb
# Added --clear option to clear refdeses.  (* jcl *)
#
# Revision 1.2  2005/12/21 00:09:56  danmc
# - Fix a bug where when using the --pgskip option, components which were
#   present on page n, but not on pages n+1 through n+j, and present again
#   on page n+j+1 got numbered in a strange way.  For example, J101 on page
#   1, no connectors on page 2, J201 on page 3 instead of J301.  Noted by
#   Stuart Brorson.
#
# - While here allow the user to change the default increment from 100 to whatever
#   they want.
#
# - Also fix a bug where exactly 101 components with the same refdes prefix
#   would cause a duplicate refdes on the next page.
#
# Revision 1.1  2003/02/21 03:21:12  ahvezda
# Added scripts/refdes_renum written by Dan McMahill
#
#
#
